File: src/gatling/scala/fr/testlab/jbp/patient/PatientSimulation.scala
package fr.testlab.jbp.patient

import io.gatling.core.Predef._
import io.gatling.http.Predef._
import scala.concurrent.duration._

// G11 B27 : simulation de charge jbp-patient
// Lance 50 utilisateurs progressivement sur 10s, chacun exécute le scénario 3 fois
// Critère de passage : responseTime.percentile95 < 500ms, succès >= 99%
class PatientSimulation extends Simulation {

  // configuration HTTP commune à tous les scénarios
  val httpProtocol = http
    .baseUrl("http://localhost:8080")            // adresse de l'app locale
    .acceptHeader("application/json")
    .header("Authorization", "Basic YWRtaW46YWRtaW4=") // admin:admin en base64

  // scénario : enchaîne GET /patients, GET /patients/1, POST /patients
  val scénario = scenario("Patients CRUD")
    .exec(
      http("GET /patients")
        .get("/api/patients")
        .check(status.is(200))
    )
    .pause(200.milliseconds)
    .exec(
      http("GET /patients/1")
        .get("/api/patients/1")
        .check(status.is(200))
    )
    .pause(200.milliseconds)
    .exec(
      http("GET /patients/999")          // id inexistant -> 404 accepté
        .get("/api/patients/999")
        .check(status.in(200, 404))
    )
    .pause(200.milliseconds)
    .exec(
      http("POST /patients")
        .post("/api/patients")
        .header("Content-Type", "application/json")
        .body(StringBody("""{"firstName":"Test","lastName":"Gatling"}"""))
        .check(status.is(201))
    )

  // injection : 50 utilisateurs injectés progressivement sur 10 secondes
  setUp(
    scénario.inject(rampUsers(50).during(10.seconds))
  ).protocols(httpProtocol)
   .assertions(
     global.responseTime.percentile(95).lt(500),  // 95e percentile < 500ms
     global.successfulRequests.percent.gte(99)     // taux de succès >= 99%
   )
}

File: src/gatling/java/fr/testlab/jbp/patient/PatientSimulation.java
package fr.testlab.jbp.patient;

import static io.gatling.javaapi.core.CoreDsl.*;
import static io.gatling.javaapi.http.HttpDsl.*;

import io.gatling.javaapi.core.*;
import io.gatling.javaapi.http.*;

import java.time.Duration;

/**
 * Simulation de charge jbp-patient — G11 B27
 *
 * Scénario : opérations CRUD sur /api/patients
 * Cible    : http://localhost:8080
 * Auth     : Basic admin:admin (YWRtaW46YWRtaW4=)
 *
 * Assertions :
 *   - P95 < 500 ms
 *   - Taux de succès >= 99 %
 *
 * Lancement : ./scripts/run-gatling.sh
 */
public class PatientSimulation extends Simulation {

    // Protocole HTTP commun à tous les scénarios
    HttpProtocolBuilder httpProtocol = http
        .baseUrl("http://localhost:8080")
        .acceptHeader("application/json")
        .contentTypeHeader("application/json")
        .header("Authorization", "Basic YWRtaW46YWRtaW4=");  // admin:admin

    // Corps JSON pour la création d'un patient
    String newPatient = "{\"firstName\":\"Gatling\",\"lastName\":\"Test\",\"birthDate\":\"2000-01-01\"}";

    // Scénario CRUD complet
    ScenarioBuilder scenario = scenario("Patients CRUD")
        // GET /api/patients — liste tous les patients
        .exec(
            http("GET /patients")
                .get("/api/patients")
                .check(status().is(200))
        )
        .pause(Duration.ofMillis(200))

        // GET /api/patients/1 — patient existant
        .exec(
            http("GET /patients/1")
                .get("/api/patients/1")
                .check(status().is(200))
        )
        .pause(Duration.ofMillis(200))

        // GET /api/patients/9999 — patient inexistant → 404
        .exec(
            http("GET /patients/9999 (not found)")
                .get("/api/patients/9999")
                .check(status().is(404))
        )
        .pause(Duration.ofMillis(200))

        // POST /api/patients — création
        .exec(
            http("POST /patients")
                .post("/api/patients")
                .body(StringBody(newPatient))
                .check(status().is(201))
                .check(jsonPath("$.id").saveAs("createdId"))
        )
        .pause(Duration.ofMillis(200))

        // GET sur le patient créé
        .exec(
            http("GET /patients/#{createdId}")
                .get("/api/patients/#{createdId}")
                .check(status().is(200))
        );

    {
        setUp(
            scenario.injectOpen(
                rampUsers(50).during(Duration.ofSeconds(10))
            )
        )
        .protocols(httpProtocol)
        .assertions(
            global().responseTime().percentile(95).lt(500),
            global().successfulRequests().percent().gte(99.0)
        );
    }
}
